crypto/cmp/cmp_msg.c

```diff
@@ -334,7 +334,7 @@ OSSL_CMP_MSG *ossl_cmp_certreq_new(OSSL_CMP_CTX *ctx, int type,
     if (type != OSSL_CMP_PKIBODY_P10CR) {
         EVP_PKEY *privkey = OSSL_CMP_CTX_get0_newPkey(ctx, 1);
 
-        if (privkey == NULL)
+        if (privkey == NULL && OSSL_CMP_CTX_get0_newPkey(ctx, 0) == NULL)
```

Comment:
I don't really understand what is going on here. Can you explain it? We seem to first ask for the public key and store it in a var call privkey. Only if that returns NULL do we then attempt to get the private key??

Reply:
Thanks for asking. The corner case fixed in this commit (with message: "ossl_cmp_certreq_new(): Fix POPO key mismatch in case newPkey is just public key") is really not easy to see and understand.

As can been seen from `doc/man3/OSSL_CMP_CTX_new.pod` or its implementation, 
`OSSL_CMP_CTX_get0_newPkey(ctx, 1)` will succeed (i.e.., return non-NULL) only if there is actually a private key, not just a public key. Any such private key will be assigned to `privkey`.
If `ctx->newPkey` contains not a private key but just a public key then we must prevent for (the usual) case of `OSSL_CRMF_POPO_SIGNATURE` that `OSSL_CMP_CTX_setup_CRM()` (called a little further down) wrongly takes this public key for inclusion in the CRM while the proof-of-possession signature would be done using a the default private key, which is the CMP signing key in `ctx->pkey`. That key would not match with the public key in `ctx->newPkey`, so we would get an invalid (non-)self-signature.
The fix is done by using the default key `ctx->pkey` only if `ctx->newPkey` does not even contain a public key, which is checked by the new condition`OSSL_CMP_CTX_get0_newPkey(ctx, 0) == NULL`.

BTW, I came across this tricky case while testing the CMP CLI with a PKCS#11-based crypto engine that (because no PIN was provided due to a configuration mistake of mine) provided just the public key portion of the key I intended to use for the cert request message.

Reply:
Thanks for the explanation. Can we put some kind of comment in the code to explain what we are doing?

Reply:
Good point - done.